### PR TITLE
[Feature] Flink Doris Connector (#5372)

### PR DESCRIPTION
**Pull request by JNSimba**
## Proposed changes

Refer to Spark connector Doris and make an example of Flink Connector Doris

1.Customize DorisSource to implement interface org.apache.flink.streaming.api.functions.source.RichSourceFunction

2.Get dorisPartitions by calling the FE execution plan Api during initialization

3.Traverse dorisPartitions, read Doris data through scalaValueReader.class

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Code refactor (Modify the code structure, format the code, etc...)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- 

- [ ]  I have created an issue on (Fix [#5371](https://github.com/apache/incubator-doris/issues/5371)) and described the bug/feature there in detail
- [ ]  Compiling and unit tests pass locally with my changes
- [ ]  I have added tests that prove my fix is effective or that my feature works
- [ ]  If these changes need document changes, I have updated the document
- [ ]  Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at dev@doris.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: extension/flink-doris-connector/src/main/java/org/apache/doris/flink/backend/BackendClient.java [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+    private Routing routing;
+
+    private TDorisExternalService.Client client;
+    private TTransport transport;
+
+    private boolean isConnected = false;
+    private final int retries;
+    private final int socketTimeout;
+    private final int connectTimeout;
+
```
Comment (pengxiangyu) [inline extension/flink-doris-connector/src/main/java/org/apache/doris/flink/backend/BackendClient.java:59]: You need to write some annotation

File: extension/flink-doris-connector/src/main/java/org/apache/doris/flink/backend/BackendClient.java [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+                logger.trace("Connect status before open transport to {} is '{}'.", routing, isConnected);
+                if (!transport.isOpen()) {
+                    transport.open();
+                    isConnected = true;
+                }
+            } catch (TTransportException e) {
+                logger.warn(ErrorMessages.CONNECT_FAILED_MESSAGE, routing, e);
+                ex = e;
+            }
+            if (isConnected) {
```
Comment (pengxiangyu) [inline extension/flink-doris-connector/src/main/java/org/apache/doris/flink/backend/BackendClient.java:92]: This is better to be in try {}(just below isConnected = true;)

File: extension/flink-doris-connector/src/main/java/org/apache/doris/flink/backend/BackendClient.java [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+                ConfigurationOptions.DORIS_REQUEST_CONNECT_TIMEOUT_MS_DEFAULT);
+        this.socketTimeout = settings.getIntegerProperty(ConfigurationOptions.DORIS_REQUEST_READ_TIMEOUT_MS,
+                ConfigurationOptions.DORIS_REQUEST_READ_TIMEOUT_MS_DEFAULT);
+        this.retries = settings.getIntegerProperty(ConfigurationOptions.DORIS_REQUEST_RETRIES,
+                ConfigurationOptions.DORIS_REQUEST_RETRIES_DEFAULT);
+        logger.trace("connect timeout set to '{}'. socket timeout set to '{}'. retries set to '{}'.",
+                this.connectTimeout, this.socketTimeout, this.retries);
+        open();
+    }
+
```
[the remark this answers is not in the document]
Comment (hf200012) [reply extension/flink-doris-connector/src/main/java/org/apache/doris/flink/backend/BackendClient.java:72]: Modification, submitted soon, and added flink SQL to create doris table

File: extension/flink-doris-connector/src/main/java/org/apache/doris/flink/backend/BackendClient.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (hf200012) [reply extension/flink-doris-connector/src/main/java/org/apache/doris/flink/backend/BackendClient.java:92]: Modification, submitted soon, and added flink SQL to create doris table

File: extension/flink-doris-connector/src/main/java/org/apache/doris/flink/backend/BackendClient.java [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+                }
+            } catch (TTransportException e) {
+                logger.warn(ErrorMessages.CONNECT_FAILED_MESSAGE, routing, e);
+                ex = e;
+            }
+            if (isConnected) {
+                logger.info("Success connect to {}.", routing);
+                break;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (hf200012) [reply extension/flink-doris-connector/src/main/java/org/apache/doris/flink/backend/BackendClient.java:96]: Modification, submitted soon, and added flink SQL to create doris table

**Comment by hf200012** [comment]
Modification, submitted soon, and added flink SQL to create doris table

**Comment by JNSimba** [comment]
Refer to [User-defined Sources & Sinks](https://ci.apache.org/projects/flink/flink-docs-stable/zh/dev/table/sourceSinks.html) And flink-connector-jdbc， add flink sql to create doris table, and write doris table through streamload。

eg:
```java
         EnvironmentSettings settings = EnvironmentSettings.newInstance()
                .useBlinkPlanner()
                .inStreamingMode()
                .build();
        TableEnvironment tEnv = TableEnvironment.create(settings);
        tEnv.getConfig().getConfiguration().setString("job.name","test");

        tEnv.executeSql(
                "CREATE TABLE doris_test (" +
                        "name STRING," +
                        "age INT," +
                        "price DECIMAL(5,2)," +
                        "sale DOUBLE" +
                        ") " +
                        "WITH (\n" +
                        "  'connector' = 'doris',\n" +
                        "  'fenodes' = '11.221.147.11:8030',\n" +
                        "  'table.identifier' = 'demo.doris_test_source_2',\n" +
                        "  'username' = 'root',\n" +
                        "  'password' = ''" +
                        ")");
        tEnv.executeSql(
                "CREATE TABLE doris_test_sink (" +
                        "name STRING," +
                        "age INT," +
                        "price DECIMAL(5,2)," +
                        "sale DOUBLE" +
                        ") " +
                        "WITH (\n" +
                        "  'connector' = 'doris',\n" +
                        "  'fenodes' = '11.221.147.11:8030',\n" +
                        "  'table.identifier' = 'demo.doris_test_sink_2',\n" +
                        "  'username' = 'root',\n" +
                        "  'password' = ''\n" +
                        ")");

        tEnv.executeSql("INSERT INTO doris_test_sink select name,age,price,sale from doris_test");
```

File: extension/flink-doris-connector/src/main/java/org/apache/doris/flink/table/DorisDynamicOutputFormat.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+                options.getFenodes(),
+                options.getTableIdentifier().split("\\.")[0],
+                options.getTableIdentifier().split("\\.")[1],
+                options.getUsername(),
+                options.getPassword());
+        System.out.println(dorisStreamLoad);
+    }
+
+    @Override
+    public void writeRecord(RowData row) throws IOException {
```
Comment (morningman) [inline extension/flink-doris-connector/src/main/java/org/apache/doris/flink/table/DorisDynamicOutputFormat.java:58]: We need to buffer a batch of rows before triggering stream load

**Comment by JNSimba** [comment]
Added batch.size and max-retries configuration items,  can set the batch size and the number of retries

File: extension/flink-doris-connector/src/main/java/org/apache/doris/flink/table/DorisDynamicOutputFormat.java (hunk 5 above)
Comment (JNSimba) [reply extension/flink-doris-connector/src/main/java/org/apache/doris/flink/table/DorisDynamicOutputFormat.java]: yes，already submitted

**Comment by JNSimba** [comment]
add build script

**Review by morningman — DISMISSED** [review]
LGTM

**Comment by EmmyMiao87** [comment]
It is recommended that you add usage methods and usage examples to the document later.

**Comment by JNSimba** [comment]
> It is recommended that you add usage methods and usage examples to the document later.

already add.

File: docs/en/extending-doris/flink-doris-connector.md [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+Execute following command in dir `extension/flink-doris-connector/`:
+
+```bash
+sh build.sh
+```
+
+After successful compilation, the file `doris-flink-1.0.0-SNAPSHOT.jar` will be generated in the `output/` directory. Copy this file to `ClassPath` in `Flink` to use `Flink-Doris-Connector`. For example, `Flink` running in `Local` mode, put this file in the `jars/` folder. `Flink` running in `Yarn` cluster mode, put this file in the pre-deployment package.
+
+## Example
```
Comment (EmmyMiao87) [inline docs/en/extending-doris/flink-doris-connector.md:50]:
```suggestion
## How to use
```

File: docs/en/extending-doris/flink-doris-connector.md [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+
+### SQL
+
```
Comment (EmmyMiao87) [inline docs/en/extending-doris/flink-doris-connector.md:53]:
```suggestion
The purpose of this step is to register the Doris data source on Flink. 
This step is operated on Flink
```

File: docs/en/extending-doris/flink-doris-connector.md [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+Execute following command in dir `extension/flink-doris-connector/`:
+
+```bash
+sh build.sh
+```
+
+After successful compilation, the file `doris-flink-1.0.0-SNAPSHOT.jar` will be generated in the `output/` directory. Copy this file to `ClassPath` in `Flink` to use `Flink-Doris-Connector`. For example, `Flink` running in `Local` mode, put this file in the `jars/` folder. `Flink` running in `Yarn` cluster mode, put this file in the pre-deployment package.
+
+## Example
+
```
Comment (EmmyMiao87) [inline docs/en/extending-doris/flink-doris-connector.md:51]:
```suggestion
The purpose of this step is to register the Doris data source on Flink. 
This step is operated on Flink.
There are two ways to use sql and scala. The following are examples to illustrate
```

File: docs/zh-CN/extending-doris/flink-doris-connector.md [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+-->
+
+# Flink Doris Connector
+
```
Comment (EmmyMiao87) [inline docs/zh-CN/extending-doris/flink-doris-connector.md:28]: 中文文档修改意见同英文

File: docs/zh-CN/extending-doris/flink-doris-connector.md [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+## 编译与安装
+
+在 `extension/flink-doris-connector/` 源码目录下执行：
+
+```bash
+sh build.sh
+```
+
+编译成功后，会在 `output/` 目录下生成文件 `doris-flink-1.0.0-SNAPSHOT.jar`。将此文件复制到 `Flink` 的 `ClassPath` 中即可使用 `Flink-Doris-Connector`。例如，`Local` 模式运行的 `Flink`，将此文件放入 `jars/` 文件夹下。`Yarn`集群模式运行的`Flink`，则将此文件放入预部署包中。
```
Comment (EmmyMiao87) [inline docs/zh-CN/extending-doris/flink-doris-connector.md:48]: 我尝试编译了一下，编译完成后发现有两个jar包。一个叫 doris-flink-1.0-SNAPSHOT.jar， 一个叫original-doris-flink-1.0-SNAPSHOT.jar。这个 original-doris-flink-1.0-SNAPSHOT.jar 是干啥用的吗？

File: extension/flink-doris-connector/src/main/java/org/apache/doris/flink/backend/BackendClient.java [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+        logger.error(ErrorMessages.CONNECT_FAILED_MESSAGE, routing);
+        throw new ConnectedFailedException(routing.toString(), ex);
+    }
+
+    /**
+     * close an scanner.
+     * @param closeParams thrift struct to required by request
+     */
+    public void closeScanner(TScanCloseParams closeParams) {
+        logger.debug("CloseScanner to '{}', parameter is '{}'.", routing, closeParams);
```
Comment (EmmyMiao87) [inline extension/flink-doris-connector/src/main/java/org/apache/doris/flink/backend/BackendClient.java:198]: CloseScanner from ? Close scanner from? CloseScanner to? Close scanner to ? Which is correct?

File: extension/flink-doris-connector/src/main/java/org/apache/doris/flink/cfg/ConfigurationOptions.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    String DORIS_READ_FIELD = "doris.read.field";
+    String DORIS_FILTER_QUERY = "doris.filter.query";
+    String DORIS_FILTER_QUERY_IN_MAX_COUNT = "doris.filter.query.in.max.count";
+    int DORIS_FILTER_QUERY_IN_VALUE_UPPER_LIMIT = 10000;
+
+    String DORIS_USER = "doris.user";
+    String DORIS_REQUEST_AUTH_USER = "doris.request.auth.user";
+    // use password to save doris.request.auth.password
+    // reuse credentials mask method in spark ExternalCatalogUtils#maskCredentials
+    String DORIS_PASSWORD = "doris.password";
```
Comment (EmmyMiao87) [inline extension/flink-doris-connector/src/main/java/org/apache/doris/flink/cfg/ConfigurationOptions.java:37]: Is there any difference between these two passwords and user?

File: docs/zh-CN/extending-doris/flink-doris-connector.md [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+### 通用配置项
+
+| Key                              | Default Value     | Comment                                                      |
+| -------------------------------- | ----------------- | ------------------------------------------------------------ |
+| fenodes                    | --                | Doris FE http 地址             |
+| table.identifier           | --                | Doris 表名，如：db1.tbl1                                 |
+| username                            | --            | 访问Doris的用户名                                            |
+| password                        | --            | 访问Doris的密码                                              |
+| sink.batch.size     | 100                | 单次写BE的最大行数        |
+| sink.max-retries     | 1                | 写BE失败之后的重试次数       |
```
Comment (EmmyMiao87) [inline docs/zh-CN/extending-doris/flink-doris-connector.md:109]: 看起来可以调整的参数好像不止这些，可以后续把参数的含义都补充一下。比如mem limit这种。

File: extension/flink-doris-connector/src/main/java/org/apache/doris/flink/rest/models/RespContent.java [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+        return Status;
+    }
+
+    public String getMessage() {
+        return Message;
+    }
+
+    @Override
+    public String toString() {
+        return "RespContent{" +
```
Comment (EmmyMiao87) [inline extension/flink-doris-connector/src/main/java/org/apache/doris/flink/rest/models/RespContent.java:84]: Json tostring is better？

File: extension/flink-doris-connector/src/test/java/org/apache/doris/flink/DorisSinkExample.java [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+
+    public static void main(String[] args) throws Exception {
+        final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+        env.setParallelism(1);
+        final StreamTableEnvironment tEnv = StreamTableEnvironment.create(env);
+        //streamload default delimiter: \t
+        DataStreamSource<String> source = env.fromElements("[{\"name\":\"doris\"}]\t1");
+        tEnv.createTemporaryView("doris_test",source,$("name"));
+
+//        Table wordWithCount = tEnv.sqlQuery("select name FROM doris_test  ");
```
Comment (EmmyMiao87) [inline extension/flink-doris-connector/src/test/java/org/apache/doris/flink/DorisSinkExample.java:35]: remove it

File: extension/flink-doris-connector/src/main/java/org/apache/doris/flink/serialization/RowBatch.java [hunk 16]
```diff
@@ -0,0 +1,304 @@
...
+            String errMsg = "Get row offset: " + rowIndex + " larger than row size: " +
+                    rowCountInOneBatch;
+            logger.error(errMsg);
+            throw new NoSuchElementException(errMsg);
+        }
+        rowBatch.get(readRowCount + rowIndex).put(obj);
+    }
+
+    public void convertArrowToRowBatch() throws DorisException {
+        try {
```
Comment (EmmyMiao87) [inline extension/flink-doris-connector/src/main/java/org/apache/doris/flink/serialization/RowBatch.java:145]: This method is worth having a unit test ~

File: extension/flink-doris-connector/src/main/java/org/apache/doris/flink/serialization/RowBatch.java [hunk 17]
```diff
@@ -0,0 +1,304 @@
...
+    }
+
+    public RowBatch(TScanBatchResult nextResult, Schema schema) throws DorisException {
+        this.schema = schema;
+        this.rootAllocator = new RootAllocator(Integer.MAX_VALUE);
+        this.arrowStreamReader = new ArrowStreamReader(
+                new ByteArrayInputStream(nextResult.getRows()),
+                rootAllocator
+                );
+        this.offsetInRowBatch = 0;
```
Comment (EmmyMiao87) [inline extension/flink-doris-connector/src/main/java/org/apache/doris/flink/serialization/RowBatch.java:97]: Although I don’t know whether it’s appropriate to put this while and exception logic in the constructor, 
but could you please changed to a more elegant way of writing.

File: extension/flink-doris-connector/src/main/java/org/apache/doris/flink/rest/models/RespContent.java (hunk 14 above)
Comment (JNSimba) [reply extension/flink-doris-connector/src/main/java/org/apache/doris/flink/rest/models/RespContent.java:84]: Modification, submitted soon

File: extension/flink-doris-connector/src/main/java/org/apache/doris/flink/serialization/RowBatch.java (hunk 16 above)
Comment (JNSimba) [reply extension/flink-doris-connector/src/main/java/org/apache/doris/flink/serialization/RowBatch.java:145]: Modification, submitted soon

File: docs/zh-CN/extending-doris/flink-doris-connector.md (hunk 13 above)
Comment (JNSimba) [reply docs/zh-CN/extending-doris/flink-doris-connector.md:109]: 参考spark-connector的参数，修改中，最近就会提交

File: docs/zh-CN/extending-doris/flink-doris-connector.md (hunk 10 above)
Comment (JNSimba) [reply docs/zh-CN/extending-doris/flink-doris-connector.md:48]: doris-flink-1.0-SNAPSHOT.jar 这个是包含flink相关依赖的包
original-doris-flink-1.0-SNAPSHOT.jar 这个是不包含依赖的包。

File: extension/flink-doris-connector/src/test/java/org/apache/doris/flink/DorisSinkExample.java (hunk 15 above)
Comment (JNSimba) [reply extension/flink-doris-connector/src/test/java/org/apache/doris/flink/DorisSinkExample.java:35]: Modification, submitted soon

File: extension/flink-doris-connector/src/main/java/org/apache/doris/flink/serialization/RowBatch.java (hunk 17 above)
Comment (JNSimba) [reply extension/flink-doris-connector/src/main/java/org/apache/doris/flink/serialization/RowBatch.java:97]: Ok, it will be modified soon

**Comment by JNSimba** [comment]
1.Merge DorisOptions and Settings configuration ;
2.Streamload is modified to BE, and retry other alive BE nodes after failure
3.add more options reference spark-connector

**Review by morningman — DISMISSED** [review]
LGTM

**Comment by JNSimba** [comment]
add pom profile patch

**Review by morningman — APPROVED** [review]
LGTM

**Comment by xhmz** [comment]
Failed to get response from Doris  http://10.0.10.28:8030/rest/v1/system?path=//backends, http code is 404

**Comment by JNSimba** [comment]
> Failed to get response from Doris http://10.0.10.28:8030/rest/v1/system?path=//backends, http code is 404

enable http v2

**Comment by xhmz** [comment]
enable_http_server_v2 ？

**Comment by JNSimba** [comment]
> enable_http_server_v2 ？

yes，enable_http_server_v2= true

**Comment by meitianjinbu** [comment]
my doris version 0.15.0 rc02，enable_http_server_v2 is true default？but failed to get response from doris http://10.0.10.28:8030/api/backends?is_alive=true

**Pull request merged**